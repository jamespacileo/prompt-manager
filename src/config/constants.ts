import path from 'path';

/**
 * Constants used throughout the Prompt Manager project.
 * This file defines important constant values that are used across different modules.
 */

/** The name of the configuration file */
export const CONFIG_FILE_NAME = process.env.PROMPT_MANAGER_CONFIG_FILE || 'prompt-manager.json';

/** The default directory name for storing prompts */
export const DEFAULT_PROMPTS_DIR = process.env.PROMPT_MANAGER_PROMPTS_DIR || 'prompts';
export const DEFAULT_CONFIG = {
  promptsDir: process.env.PROMPT_MANAGER_PROMPTS_DIR || 'prompts',
  outputDir: process.env.PROMPT_MANAGER_OUTPUT_DIR || 'generated',
  preferredModels: process.env.PROMPT_MANAGER_PREFERRED_MODELS ? JSON.parse(process.env.PROMPT_MANAGER_PREFERRED_MODELS) : ['gpt-4', 'gpt-3.5-turbo'],
  modelParams: process.env.PROMPT_MANAGER_MODEL_PARAMS ? JSON.parse(process.env.PROMPT_MANAGER_MODEL_PARAMS) : {
    'gpt-4': {
      temperature: 0.7,
      maxTokens: 2000,
      topP: 1,
      frequencyPenalty: 0,
      presencePenalty: 0
    },
    'gpt-3.5-turbo': {
      temperature: 0.8,
      maxTokens: 1500,
      topP: 1,
      frequencyPenalty: 0,
      presencePenalty: 0
    }
  },
  defaultLanguage: process.env.PROMPT_MANAGER_DEFAULT_LANGUAGE || 'en',
  version: process.env.PROMPT_MANAGER_VERSION || '1.0.0'
};

export function getProjectRoot(): string {
  return process.env.PROMPT_MANAGER_PROJECT_ROOT || process.env.FURY_PROJECT_ROOT || process.cwd();
}

export function getConfigPath(): string {
  return process.env.PROMPT_MANAGER_CONFIG_PATH || path.resolve(getProjectRoot(), CONFIG_FILE_NAME);
}

export function getDefaultPromptsPath(): string {
  return process.env.PROMPT_MANAGER_DEFAULT_PROMPTS_PATH || path.resolve(getProjectRoot(), DEFAULT_PROMPTS_DIR);
}
