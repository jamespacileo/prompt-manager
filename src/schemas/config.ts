import { z } from 'zod';

export const configSchema = z.object({
  promptsDir: z.string().describe('Directory for storing prompts'),
  outputDir: z.string().describe('Directory for storing output'),
  preferredModels: z.array(z.string()).describe('List of preferred AI models'),
  modelParams: z.record(z.object({
    temperature: z.number().min(0).max(1).optional().describe('AI model temperature parameter'),
    maxTokens: z.number().positive().optional().describe('Maximum number of tokens for AI model output'),
    topP: z.number().min(0).max(1).optional().describe('Top P parameter for AI model'),
    frequencyPenalty: z.number().min(-2).max(2).optional().describe('Frequency penalty for AI model'),
    presencePenalty: z.number().min(-2).max(2).optional().describe('Presence penalty for AI model'),
  })).describe('Parameters for different AI models'),
});

export type Config = z.infer<typeof configSchema>;

export const DEFAULT_CONFIG: Config = {
  promptsDir: '',
  outputDir: '',
  preferredModels: ['gpt-4', 'gpt-4o-mini'],
  modelParams: {
    'gpt-4': {
      temperature: 0.7,
      maxTokens: 2000,
      topP: 1,
      frequencyPenalty: 0,
      presencePenalty: 0,
    },
    'gpt-4o-mini': {
      temperature: 0.8,
      maxTokens: 1500,
      topP: 1,
      frequencyPenalty: 0,
      presencePenalty: 0,
    },
  },
};
