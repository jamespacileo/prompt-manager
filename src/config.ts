import { Container } from "typedi";
import { PromptProjectConfigManager } from "./config/PromptProjectConfigManager";
import type { Config } from "./schemas/config";

export async function initializeConfig(): Promise<void> {
	const configManager = Container.get(PromptProjectConfigManager);
	await configManager.initialize();
}

export async function getConfig<K extends keyof Config>(
	key: K,
): Promise<Config[K]> {
	const configManager = Container.get(PromptProjectConfigManager);
	return configManager.getConfig(key);
}

export async function getAllConfig(): Promise<Config> {
	const configManager = Container.get(PromptProjectConfigManager);
	return configManager.getAllConfig();
}

export async function updateConfig(updates: Partial<Config>): Promise<void> {
	const configManager = Container.get(PromptProjectConfigManager);
	await configManager.updateConfig(updates);
}

export type { Config };
