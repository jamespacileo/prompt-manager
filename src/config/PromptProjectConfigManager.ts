import { cosmiconfig, type OptionsSync } from "cosmiconfig";
import path from "path";
import type { IPromptProjectConfigManager } from "../types/interfaces";
import { ensureDirectoryExists } from "../utils/fileUtils";
import { configSchema, DEFAULT_CONFIG, z } from "../schemas/config";
import type { Config } from "../schemas/config";
import { Service } from "typedi";
import { fromError } from "zod-validation-error";
import { logger } from "../utils/logger";
import chalk from "chalk";

const DEFAULT_PROMPTS_FOLDER_CONFIG_FILENAME = ".promptmanager.config.json";

export type { Config };

@Service()
export class PromptProjectConfigManager implements IPromptProjectConfigManager {
	private config: Config;
	private initialized = false;
	private explorer;
	private basePath: string;
	private configFilePath: string | undefined;

	constructor() {
		this.config = { ...DEFAULT_CONFIG };
		this.basePath = process.env.FURY_PROJECT_ROOT || process.cwd();
		const explorerOptions: OptionsSync = {
			searchPlaces: [
				"package.json",
				".promptmanagerrc",
				".promptmanagerrc.json",
				".promptmanagerrc.yaml",
				".promptmanagerrc.yml",
				".promptmanagerrc.js",
				"promptmanager.config.js",
			],
			loaders: {},
			transform: (result) => result,
			ignoreEmptySearchPlaces: true,
			cache: true,
			mergeImportArrays: true,
			mergeSearchPlaces: true,
			searchStrategy: "project",
		};
		this.explorer = cosmiconfig("promptmanager", explorerOptions);
		logger.debug(
			`Initializing PromptProjectConfigManager with basePath: ${this.basePath}`,
		);
	}

	public getBasePath(): string {
		return this.basePath;
	}

	public getPromptsDir(): string {
		if (!this.config.promptsDir) {
			throw new Error(
				"Prompts directory not set. Please set the prompts directory in your configuration file.",
			);
		}
		if (!path.isAbsolute(this.config.promptsDir)) {
			return path.join(this.basePath, this.config.promptsDir);
		}
		return this.config.promptsDir;
	}

	public getProjectConfigPath(): string {
		if (!this.configFilePath) {
			logger.warn("No configuration file found. Using default configuration.");
			throw new Error(
				"No configuration file found. Please set the configuration file in your project.",
			);
		}
		return this.configFilePath;
	}

	public async initialize(): Promise<void> {
		if (this.initialized) return;

		try {
			await this.loadConfig();
			await this.ensureConfigDirectories();
			this.prettyPrintConfig();
			this.initialized = true;
		} catch (error: any) {
			logger.error("Failed to initialize config:", error.message, error.stack);
			throw new Error(
				"Failed to initialize PromptProjectConfigManager. Please check your configuration and try again.",
				error.stack,
			);
		}
	}

	public async isInitialized(): Promise<boolean> {
		return this.initialized;
	}

	private async loadConfig(): Promise<void> {
		try {
			const result = await this.explorer.search();
			if (result && result.config) {
				const validatedConfig = configSchema.parse({
					...DEFAULT_CONFIG,
					...result.config,
				});
				this.config = {
					...validatedConfig,
					promptsDir: path.resolve(this.basePath, validatedConfig.promptsDir),
					outputDir: path.resolve(this.basePath, validatedConfig.outputDir),
				};
				logger.debug(`Found configuration file at ${result.filepath}`);
				this.configFilePath = result.filepath;
			} else {
				logger.warn(
					"No configuration file found. Using default configuration.",
				);
				this.config = { ...DEFAULT_CONFIG };
			}
			if (!this.configFilePath) {
				logger.error(`No configuration file found at ${this.basePath}`);
				this.configFilePath = path.join(
					this.basePath,
					DEFAULT_PROMPTS_FOLDER_CONFIG_FILENAME,
				);
			} else {
				logger.debug(`Configuration file found at ${this.configFilePath}`);
			}
			logger.debug("Configuration loaded successfully");
			logger.debug(`Prompts directory: ${this.config.promptsDir}`);
			logger.debug(`Output directory: ${this.config.outputDir}`);
		} catch (error) {
			if (error instanceof z.ZodError) {
				const validationError = fromError(error);
				logger.error(validationError.toString());
				logger.error("Invalid configuration:", error.errors);
				throw new Error(`Invalid configuration: ${error.message}`);
			} else {
				logger.error("Error loading configuration:", error);
				throw new Error(
					"Failed to load configuration. Please check your configuration file.",
				);
			}
		}
	}

	private async ensureConfigDirectories(): Promise<void> {
		try {
			await ensureDirectoryExists(this.config.promptsDir);
			logger.success(
				`Created prompts directory: ${chalk.yellow(this.config.promptsDir)}`,
			);
			await ensureDirectoryExists(this.config.outputDir);
			logger.success(
				`Created output directory: ${chalk.yellow(this.config.outputDir)}`,
			);
		} catch (error: any) {
			logger.error("Error: Failed to create necessary directories.");
			logger.warn(
				"Please check that you have write permissions in the project directory.",
			);
			throw new Error(`Failed to create directories: ${error.message}`);
		}
	}

	private prettyPrintConfig(): void {
		if (this.config.verbosity > 99) {
			logger.info(chalk.bold("\nLoaded Configuration:"));
			logger.info(`prompts_dir:      ${chalk.cyan(this.config.promptsDir)}`);
			logger.info(`output_dir:       ${chalk.cyan(this.config.outputDir)}`);
			logger.info(
				`preferredModels:  ${chalk.cyan(this.config.preferredModels.join(", "))}`,
			);
			logger.info("modelParams:");
			Object.entries(this.config.modelParams).forEach(([model, params]) => {
				logger.info(`  ${model}:`);
				Object.entries(params).forEach(([key, value]) => {
					logger.info(`    ${key}: ${chalk.cyan(value)}`);
				});
			});
			logger.info(`verbosity:        ${chalk.cyan(this.config.verbosity)}`);
			logger.info("\n");
		}
	}

	public getConfig<K extends keyof Config>(key: K): Config[K] {
		return this.config[key];
	}

	public getAllConfig(): Config {
		return { ...this.config };
	}

	public async updateConfig(newConfig: Partial<Config>): Promise<void> {
		try {
			const updatedConfig = configSchema.parse({
				...this.config,
				...newConfig,
			});
			this.config = updatedConfig;
			this.prettyPrintConfig();
		} catch (error: any) {
			if (error instanceof z.ZodError) {
				const validationError = fromError(error);
				logger.error("Validation error:", validationError.toString());
				logger.error("Invalid configuration update:", error.errors);
				throw new Error(`Invalid configuration update: ${error.message}`);
			}
			throw error;
		}
	}
	setVerbosity(level: number): void {
		this.config.verbosity = level;
	}

	getVerbosity(): number {
		return this.config.verbosity;
	}
}
