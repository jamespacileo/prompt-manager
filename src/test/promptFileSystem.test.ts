import {
	expect,
	test,
	beforeAll,
	afterAll,
	beforeEach,
	afterEach,
	describe,
} from "bun:test";
import { Container } from "typedi";
import { PromptFileSystem } from "../promptFileSystem";
import { PromptProjectConfigManager } from "../config/PromptProjectConfigManager";
import type { IPrompt, IPromptInput, IPromptOutput } from "../types/interfaces";
import fs from "fs/promises";
import path from "path";

const COSMIC_PROMPT: IPrompt<IPromptInput, IPromptOutput> = {
	name: "cosmicWhisper",
	category: "celestialMystery",
	description: "A prompt that echoes the secrets of the universe",
	version: "1.0.0",
	template:
		"In the vast expanse of cosmic wonder, {{celestialBody}} reveals its hidden truths",
	parameters: ["celestialBody"],
	metadata: {
		created: new Date().toISOString(),
		lastModified: new Date().toISOString(),
	},
	outputSchema: {
		type: "object",
		properties: {
			cosmicTruth: { type: "string" },
		},
		required: ["cosmicTruth"],
	},
	outputType: "structured",
	inputSchema: {
		type: "object",
		properties: {
			celestialBody: { type: "string" },
		},
		required: ["celestialBody"],
	},
	configuration: {
		modelName: "default-model",
		temperature: 0.7,
		maxTokens: 100,
		topP: 1,
		frequencyPenalty: 0,
		presencePenalty: 0,
		stopSequences: [],
	},
};

describe("PromptFileSystem", () => {
	let promptFileSystem: PromptFileSystem;
	let testDir: string;
	let originalPromptsDir: string;
	let configManager: PromptProjectConfigManager;

	beforeAll(async () => {
		originalPromptsDir = process.env.PROMPTS_DIR || "";
		testDir = path.resolve(process.cwd(), "test-prompts");
		await fs.mkdir(testDir, { recursive: true });
		process.env.PROMPTS_DIR = testDir;
	});

	beforeEach(async () => {
		Container.reset();

		configManager = Container.get(PromptProjectConfigManager);
		await configManager.initialize();
		await configManager.updateConfig({ promptsDir: testDir });

		promptFileSystem = Container.get(PromptFileSystem);
		await promptFileSystem.initialize();
	});

	afterEach(async () => {
		await fs.rm(testDir, { recursive: true, force: true });
		await fs.mkdir(testDir, { recursive: true });
	});

	afterAll(async () => {
		await fs.rm(testDir, { recursive: true, force: true });
		if (originalPromptsDir) {
			process.env.PROMPTS_DIR = originalPromptsDir;
		} else {
			delete process.env.PROMPTS_DIR;
		}
	});

	test("savePrompt and loadPrompt", async () => {
		await promptFileSystem.savePrompt({ promptData: COSMIC_PROMPT });
		const loadedPrompt = await promptFileSystem.loadPrompt({
			category: "celestialMystery",
			promptName: "cosmicWhisper",
		});

		expect(loadedPrompt).toEqual(COSMIC_PROMPT);
	});

	test("promptExists", async () => {
		await promptFileSystem.savePrompt({ promptData: COSMIC_PROMPT });
		const exists = await promptFileSystem.promptExists({
			category: "celestialMystery",
			promptName: "cosmicWhisper",
		});
		expect(exists).toBe(true);

		const notExists = await promptFileSystem.promptExists({
			category: "celestialMystery",
			promptName: "nonExistentPrompt",
		});
		expect(notExists).toBe(false);
	});

	test("listPrompts", async () => {
		await promptFileSystem.savePrompt({ promptData: COSMIC_PROMPT });
		const stargazerPrompt = {
			...COSMIC_PROMPT,
			name: "stargazerDreams",
			description: "A prompt that captures the essence of stargazing",
		};
		await promptFileSystem.savePrompt({ promptData: stargazerPrompt });

		const prompts = await promptFileSystem.listPrompts();
		expect(prompts.length).toBeGreaterThanOrEqual(2);
		// expect(prompts).toContainEqual(
		//   expect.objectContaining({
		//     name: "cosmicWhisper",
		//     category: "celestialMystery",
		//     filePath: expect.stringContaining(path.join("celestialMystery", "cosmicWhisper", "prompt.json"))
		//   })
		// );
		// expect(prompts).toContainEqual(
		//   expect.objectContaining({
		//     name: "stargazerDreams",
		//     category: "celestialMystery",
		//     filePath: expect.stringContaining(path.join("celestialMystery", "stargazerDreams", "prompt.json"))
		//   })
		// );

		const categoryPrompts = await promptFileSystem.listPrompts({
			category: "celestialMystery",
		});
		expect(categoryPrompts).toMatchSnapshot();
		expect(categoryPrompts.length).toBeGreaterThanOrEqual(2);
	});
});
