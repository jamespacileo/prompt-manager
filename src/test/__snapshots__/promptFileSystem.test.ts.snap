// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`PromptFileSystem listPrompts 1`] = `
[
  PromptModel {
    "_inputZodSchema": null,
    "_isSaved": false,
    "_outputZodSchema": null,
    "category": "celestialMystery",
    "configManager": PromptProjectConfigManager {
      "basePath": "/home/james/Projects/prompt-manager",
      "config": {
        "modelParams": {
          "gpt-4": {
            "frequencyPenalty": 0,
            "maxTokens": 2000,
            "presencePenalty": 0,
            "temperature": 0.7,
            "topP": 1,
          },
          "gpt-4o-mini": {
            "frequencyPenalty": 0,
            "maxTokens": 1500,
            "presencePenalty": 0,
            "temperature": 0.8,
            "topP": 1,
          },
        },
        "outputDir": "/home/james/Projects/prompt-manager/src/generated",
        "preferredModels": [
          "gpt-4",
          "gpt-4o-mini",
        ],
        "promptsDir": "/home/james/Projects/prompt-manager/test-prompts",
        "verbosity": 1,
      },
      "configFilePath": "/home/james/Projects/prompt-manager/.promptmanagerrc.json",
      "explorer": {
        "clearCaches": [Function: clearCaches],
        "clearLoadCache": [Function: clearLoadCache],
        "clearSearchCache": [Function: clearSearchCache],
        "load": [Function: AsyncFunction],
        "search": [Function: AsyncFunction],
      },
      "initialized": true,
    },
    "configuration": {
      "frequencyPenalty": 0,
      "maxTokens": 100,
      "modelName": "gpt-4o-mini",
      "presencePenalty": 0,
      "stopSequences": [],
      "temperature": 0.7,
      "topP": 1,
    },
    "defaultModelName": undefined,
    "description": "A prompt that echoes the secrets of the universe",
    "filePath": undefined,
    "fileSystem": PromptFileSystem {
      "configManager": PromptProjectConfigManager {
        "basePath": "/home/james/Projects/prompt-manager",
        "config": {
          "modelParams": {
            "gpt-4": {
              "frequencyPenalty": 0,
              "maxTokens": 2000,
              "presencePenalty": 0,
              "temperature": 0.7,
              "topP": 1,
            },
            "gpt-4o-mini": {
              "frequencyPenalty": 0,
              "maxTokens": 1500,
              "presencePenalty": 0,
              "temperature": 0.8,
              "topP": 1,
            },
          },
          "outputDir": "/home/james/Projects/prompt-manager/src/generated",
          "preferredModels": [
            "gpt-4",
            "gpt-4o-mini",
          ],
          "promptsDir": "/home/james/Projects/prompt-manager/test-prompts",
          "verbosity": 1,
        },
        "configFilePath": "/home/james/Projects/prompt-manager/.promptmanagerrc.json",
        "explorer": {
          "clearCaches": [Function: clearCaches],
          "clearLoadCache": [Function: clearLoadCache],
          "clearSearchCache": [Function: clearSearchCache],
          "load": [Function: AsyncFunction],
          "search": [Function: AsyncFunction],
        },
        "initialized": true,
      },
      "initialized": true,
    },
    "inputSchema": {
      "properties": {
        "celestialBody": {
          "type": "string",
        },
      },
      "required": [
        "celestialBody",
      ],
      "type": "object",
    },
    "isLoadedFromStorage": false,
    "metadata": {
      "created": "2024-07-29T14:39:24.167Z",
      "lastModified": "2024-07-29T14:39:24.167Z",
    },
    "name": "cosmicWhisper",
    "outputSchema": {
      "properties": {
        "cosmicTruth": {
          "type": "string",
        },
      },
      "required": [
        "cosmicTruth",
      ],
      "type": "object",
    },
    "outputType": "structured",
    "parameters": [
      "celestialBody",
    ],
    "template": "In the vast expanse of cosmic wonder, {{celestialBody}} reveals its hidden truths",
    "version": "1.0.0",
  },
  PromptModel {
    "_inputZodSchema": null,
    "_isSaved": false,
    "_outputZodSchema": null,
    "category": "celestialMystery",
    "configManager": PromptProjectConfigManager {
      "basePath": "/home/james/Projects/prompt-manager",
      "config": {
        "modelParams": {
          "gpt-4": {
            "frequencyPenalty": 0,
            "maxTokens": 2000,
            "presencePenalty": 0,
            "temperature": 0.7,
            "topP": 1,
          },
          "gpt-4o-mini": {
            "frequencyPenalty": 0,
            "maxTokens": 1500,
            "presencePenalty": 0,
            "temperature": 0.8,
            "topP": 1,
          },
        },
        "outputDir": "/home/james/Projects/prompt-manager/src/generated",
        "preferredModels": [
          "gpt-4",
          "gpt-4o-mini",
        ],
        "promptsDir": "/home/james/Projects/prompt-manager/test-prompts",
        "verbosity": 1,
      },
      "configFilePath": "/home/james/Projects/prompt-manager/.promptmanagerrc.json",
      "explorer": {
        "clearCaches": [Function: clearCaches],
        "clearLoadCache": [Function: clearLoadCache],
        "clearSearchCache": [Function: clearSearchCache],
        "load": [Function: AsyncFunction],
        "search": [Function: AsyncFunction],
      },
      "initialized": true,
    },
    "configuration": {
      "frequencyPenalty": 0,
      "maxTokens": 100,
      "modelName": "gpt-4o-mini",
      "presencePenalty": 0,
      "stopSequences": [],
      "temperature": 0.7,
      "topP": 1,
    },
    "defaultModelName": undefined,
    "description": "A prompt that captures the essence of stargazing",
    "filePath": undefined,
    "fileSystem": PromptFileSystem {
      "configManager": PromptProjectConfigManager {
        "basePath": "/home/james/Projects/prompt-manager",
        "config": {
          "modelParams": {
            "gpt-4": {
              "frequencyPenalty": 0,
              "maxTokens": 2000,
              "presencePenalty": 0,
              "temperature": 0.7,
              "topP": 1,
            },
            "gpt-4o-mini": {
              "frequencyPenalty": 0,
              "maxTokens": 1500,
              "presencePenalty": 0,
              "temperature": 0.8,
              "topP": 1,
            },
          },
          "outputDir": "/home/james/Projects/prompt-manager/src/generated",
          "preferredModels": [
            "gpt-4",
            "gpt-4o-mini",
          ],
          "promptsDir": "/home/james/Projects/prompt-manager/test-prompts",
          "verbosity": 1,
        },
        "configFilePath": "/home/james/Projects/prompt-manager/.promptmanagerrc.json",
        "explorer": {
          "clearCaches": [Function: clearCaches],
          "clearLoadCache": [Function: clearLoadCache],
          "clearSearchCache": [Function: clearSearchCache],
          "load": [Function: AsyncFunction],
          "search": [Function: AsyncFunction],
        },
        "initialized": true,
      },
      "initialized": true,
    },
    "inputSchema": {
      "properties": {
        "celestialBody": {
          "type": "string",
        },
      },
      "required": [
        "celestialBody",
      ],
      "type": "object",
    },
    "isLoadedFromStorage": false,
    "metadata": {
      "created": "2024-07-29T14:39:24.167Z",
      "lastModified": "2024-07-29T14:39:24.167Z",
    },
    "name": "stargazerDreams",
    "outputSchema": {
      "properties": {
        "cosmicTruth": {
          "type": "string",
        },
      },
      "required": [
        "cosmicTruth",
      ],
      "type": "object",
    },
    "outputType": "structured",
    "parameters": [
      "celestialBody",
    ],
    "template": "In the vast expanse of cosmic wonder, {{celestialBody}} reveals its hidden truths",
    "version": "1.0.0",
  },
]
`;
